src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -564,14 +566,17 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 				return this.connectWithOptions(connection, uri, options, callbacks);
 			}
 			else {
-				let connectionErrorHandled = await this._errorDiagnosticsService.tryHandleConnectionError(connectionResult.errorCode, connectionResult.errorMessage, connection.providerName, connection);
-				if (connectionErrorHandled.handled) {
+				let connectionErrorHandleResult = await this._errorDiagnosticsService.tryHandleConnectionError(connectionResult, connection.providerName, connection);
+				if (connectionErrorHandleResult.handled) {
 					connectionResult.errorHandled = true;
-					if (connectionErrorHandled.options) {
+					if (connectionErrorHandleResult.options) {
 						//copy over altered connection options from the result if provided.
-						connection.options = connectionErrorHandled.options;
+						connection.options = connectionErrorHandleResult.options;
 					}
-					return this.connectWithOptions(connection, uri, options, callbacks);
+					// Attempt reconnect if requested by provider
+					return connectionErrorHandleResult.reconnect
```

Comment:
So currently you have your code passing handled as true but then reconnect as false if the user exits out of the TSC dialog. That means that the connection will have failed but won't result in the onConnectReject callback being called. 

If reconnect is false shouldn't we always call onConnectReject? Since the user didn't change anything about the existing connection so it still has to be in a failed state.

Reply:
The connection has already failed, and since the error dialog is already shown, state does not change. 
Calling 'onConnectReject' leads to showing the error dialog again, which is not what we want when the error dialog is already shown.

Reply:
Make sure you look at other uses of `onConnectReject`. For example I believe the query editor uses that to reset the "connecting" state

https://github.com/Microsoft/azuredatastudio/blob/8a3d08f0dec1cdf47d7fc442f90fb2227b828375/src/sql/workbench/common/editor/query/queryEditorInput.ts#L290

and OE uses that to reset the loading spinner

https://github.com/Microsoft/azuredatastudio/blob/83173c4ce536c74fb24a92c28b7060b601aa7c54/src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts#L186

Reply:
Addressed this with onConnectCancel callback